topology.py

```diff
@@ -1,42 +1,164 @@
-        # and port via etcd
-        self.nodes_metadata_store.pop(node_id)
-        self.nodes_pipeline.remove(node_id)
+        """
+        Removes a node and all its metadata from the topology. Also updates the next_node for its previous node/prev_node for the next node.
+        """
+
+        # updates the "linked list" in the prev_node and next_node metadata in etcd.
+        try:
+            prev_node = self.etcd_client.get(f"nodes/{node_id}/prev_node")
+            next_node = self.etcd_client.get(f"nodes/{node_id}/next_node")
+            
+            prev_node_id = int(prev_node[0]) if prev_node[0] else -1
+            next_node_id = int(next_node[0]) if next_node[0] else -1
+
+            if prev_node_id > -1:
+                self.etcd_client.put(f"nodes/{prev_node_id}/next_node", str(next_node))
+
+            if next_node_id > -1:
+                self.etcd_client.put(f"nodes/{next_node_id}/prev_node", str(prev_node))
 
+            # removes node from pipeline and etcd
+            self.nodes_pipeline.remove(node_id)
+            self.etcd_client.delete(f"nodes/{node_id}",dir=True)
+
+            self.logger.info("Deleted node %s from topology and memory.", node_id)
+
+        except Exception as e:
+            self.logger.error("Unable to remove node %s: %s", node_id, e)
+
+        
     def __str__(self):
         ret_str = "Topology:\n"
-        for k, v in self.nodes_metadata_store.items():
-            ret_str += f"Node ID: {k}, IP Address: {v.ip_addr}, Port: {v.port}\n"
+        for node_id in self.nodes_pipeline:
+            node_metadata = self.get_node_metadata(node_id)
+            if not node_metadata:
+                ret_str += f"Node ID: {node_id}: Error getting metadata\n"
+                continue
+            ret_str += f"Node ID: {node_id}, IP Address: {node_metadata.get(etcd_utils.IP.ADDR, "")}, Port: {node_metadata.get(etcd_utils.IP.PORT, "")}\n"
+        
         return ret_str
+        
+
+    def get_node_metadata(self, node_id):
+        """
+        Returns node metadata as a dict if it exists. If it doesn't, returns None.
+        """
+        metadata, _ = self.etcd_client.get(f"nodes/{node_id}/metadata")
+        if metadata:
+            return json.loads(metadata.decode('utf-8'))
+        return None
+    
+    # def broadcast_new_topology(self, node_id):
+    #     for k, v in self.nodes_metadata_store.items():
+    #         if k == node_id:
+    #             continue
+    #         print(
+    #             f"[broadcast_new_topology] Node ID: {k}, IP Address: {v.get_ip_addr()}, Port: {v.get_port()}"
+    #         )
+    #     for node_id in self.nodes_topology:
+    #         # self.nodes_metadata_store[node_id].client.membership_change(self)
+    #         # TODO: handle topology change
+    #         print(node_id)
+
+t = Topology("localhost", 64629, "localhost")
+t.append(0)
+t.append(1)
+# print(t.get_next_node_id(0))
+print(t)
```

Comment:
I see the confusion now. So here's what the desired behaviour should look like
- Start coordinator `coordinator.py` 
- Start nodes `node.py` (either explicitly pass in port or use your helper function to grab an available port) 
- Upon startup, nodes will call NodeJoin and there should be no errors
- Disconnect some nodes (ctrl c or wtv signal). Upon disconnected, nodes should call NodeLeave, should be no error here
- Between any of these steps, etcd meta store should topology information

Reply:
Some open design question for you to think about. Since we already moved Topology information to etcd, we can consider doing the followings

- We might want to change return type of NodeJoin to Void and let node fetch metadata of its neighbor, either sync or async. Take global sync into account when deciding whether to do this sync or async.
- Keep return type of NodeJoin to Void and periodically fetch metadata of neighbors